vm/interpreter/src/default_syscalls.rs

```diff
@@ -0,0 +1,153 @@
+        // (which runs a syntactic check), we do it directly on the CIDs.
+
+        // (0) cheap preliminary checks
+
+        if h1 == h2 {
+            return Err(format!(
+                "no consensus fault: submitted blocks are the same: {:?}, {:?}",
+                h1, h2
+            )
+            .into());
+        };
+        let bh_1: BlockHeader = from_slice(h1)?;
+        let bh_2: BlockHeader = from_slice(h2)?;
+
+        // (1) check conditions necessary to any consensus fault
+
+        if bh_1.miner_address() != bh_2.miner_address() {
+            return Err(format!(
+                "no consensus fault: blocks not mined by same miner: {:?}, {:?}",
+                bh_1.miner_address(),
+                bh_2.miner_address()
+            )
+            .into());
+        };
+        // block a must be earlier or equal to block b, epoch wise (ie at least as early in the chain).
+        if bh_1.epoch() < bh_2.epoch() {
+            return Err(format!(
+                "first block must not be of higher height than second: {:?}, {:?}",
+                bh_1.epoch(),
+                bh_2.epoch()
+            )
+            .into());
+        };
+        let mut cf: Option<ConsensusFault> = None;
+        // (a) double-fork mining fault
+        if bh_1.epoch() == bh_2.epoch() {
+            cf = Some(ConsensusFault {
+                target: *bh_1.miner_address(),
+                epoch: bh_2.epoch(),
+                fault_type: ConsensusFaultType::DoubleForkMining,
+            })
+        };
+        // (b) time-offset mining fault
+        // strictly speaking no need to compare heights based on double fork mining check above,
+        // but at same height this would be a different fault.
+        if bh_1.parents() != bh_2.parents() && bh_1.epoch() != bh_2.epoch() {
+            cf = Some(ConsensusFault {
+                target: *bh_1.miner_address(),
+                epoch: bh_2.epoch(),
+                fault_type: ConsensusFaultType::TimeOffsetMining,
+            })
+        };
+        // (c) parent-grinding fault
+        // Here extra is the "witness", a third block that shows the connection between A and B as
+        // A's sibling and B's parent.
+        // Specifically, since A is of lower height, it must be that B was mined omitting A from its tipset
+        if !extra.is_empty() {
+            let bh_3: BlockHeader = from_slice(extra)?;
+            if bh_1.parents() != bh_3.parents()
+                && bh_1.epoch() != bh_3.epoch()
+                && bh_2.parents().cids().contains(bh_3.cid())
+                && !bh_2.parents().cids().contains(bh_1.cid())
+            {
+                cf = Some(ConsensusFault {
+                    target: *bh_1.miner_address(),
+                    epoch: bh_2.epoch(),
+                    fault_type: ConsensusFaultType::ParentGrinding,
+                })
+            }
+        };
```

Comment:
In Lotus, they have a step (3) where they check at this point of consensusFault == nil. Do we not need that?

Reply:
That is achieved by cf being an Option